File: asgn4/all_nearest_neighbors_kdtree.cpp
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <algorithm>
#include <cmath>
#include <queue>
#include <fstream>
#include <string>
#include <sstream>
using namespace std;

#define DIM 3
#define NUM_NEAREST_NEEDED 3
#define NUM_POINTS 1000000

struct point{
  double coordinates[DIM];
};

struct point_distance{
  point p;
  double distance;

  bool operator< (const point_distance& a) const {
    return distance<a.distance;
  }
};

priority_queue<point_distance> knearest;

class kdtree {
public:
  class compare_points{
    int axis;
  public:
    compare_points(int d){
      axis = d;
    }

    bool operator() (point a, point b){
      return a.coordinates[axis]<b.coordinates[axis];
    }
  };

  point location;
  int axis;
  kdtree* left_child;
  kdtree* right_child;  
  point_distance tmp;

  double sumaverage;  

  kdtree(point* point_list, int num_points, int depth) {
    sumaverage = 0.0;    
    if (num_points==0) {
      this->axis=-1;
      this->left_child = NULL;
      this->right_child = NULL;
    } else {      
      this->axis = depth % DIM;
      sort(point_list, point_list+num_points, compare_points(this->axis));
      int median = num_points/2;
      
      this->location = point_list[median];
      this->left_child = new kdtree(point_list, median, depth+1);
      this->right_child = new kdtree(point_list+median+1, num_points-median-1, depth+1);      
    }
  }

  void k_nearest_neighbors(point query) {
    if (this->axis==-1)
      return;
    else {     
      double cur_distance = distance(this->location, query);
      if (knearest.size() < NUM_NEAREST_NEEDED+1) {	
	tmp.p = this->location;
	tmp.distance = cur_distance;
	knearest.push(tmp);
      }      
      else if (cur_distance < knearest.top().distance) {
	knearest.pop();       
	tmp.p = this->location;
	tmp.distance = cur_distance;
	knearest.push(tmp);
      }
      
      kdtree* nc = near_child(query);
      nc->k_nearest_neighbors(query);

      double max_dist = knearest.top().distance;
      if (knearest.size()<NUM_NEAREST_NEEDED+1 || axis_distance(query) < max_dist) {
	kdtree* fc = far_child(query);
	fc->k_nearest_neighbors(query);
      }
      return;
    }
  }   

  void print_knearest(point query) {
    while(!knearest.empty()){
      knearest.pop();
    }    
    k_nearest_neighbors(query);
    int total = NUM_NEAREST_NEEDED;
    double sum = 0.0;
    while (total>0) {      
      sum += knearest.top().distance;
      knearest.pop();
      total--;
    }
    sumaverage += sum/NUM_NEAREST_NEEDED;
  }

  kdtree* near_child(point p) {
    if (p.coordinates[this->axis] < this->location.coordinates[this->axis])
      return this->left_child;
    else
      return this->right_child;
  }

  kdtree* far_child(point p) {
    if (p.coordinates[this->axis] < this->location.coordinates[this->axis])
      return this->right_child;
    else
      return this->left_child;
  }

  double axis_distance(point p) {
    double tmp = fabs(this->location.coordinates[this->axis] - p.coordinates[this->axis]);
    return tmp;
    // return tmp*tmp;
  }

  double distance(point a, point b) {
    double dist = 0.0;
    for (int i=0; i<DIM; i++) {
      double tmp = a.coordinates[i]-b.coordinates[i];
      dist += tmp*tmp;
    }
    return sqrt(dist);
  }
};

point parse_string_as_point(string input){
  point ret;
  istringstream iss(input);
  string n;
  int i=0;
  while(getline(iss, n, ',')) { 
    ret.coordinates[i] = atof(n.c_str());     
    i++;
  }  
  return ret;
}

int main(){
  string line;
  string input_file = "nkatariy.dat";
  ifstream file(input_file.c_str());
  if (file.is_open()) {
    getline(file, line);
    file.close();

    point* point_list = new point[NUM_POINTS];
    point* query_points = new point[NUM_POINTS];
    stringstream ss(line);
    string token;
    for(int i=0; i<NUM_POINTS; i++) {
      getline(ss, token, ';');
      point p = parse_string_as_point(token);
      point_list[i] = p;
      query_points[i] = p;
    }   
    // printf("Created point list\n");
    kdtree kdt(point_list, NUM_POINTS, 0);
    // printf("Built kdtree\n");
    for (int i=0; i<NUM_POINTS; i++){
      kdt.print_knearest(query_points[i]);
    }
    printf("Average distance: %lf\n", kdt.sumaverage);
  } else{
    cout<<"Could not open file"<<endl;
  }  
  
  return 0;
}

File: asgn4/all_nearest_neighbors_kdtree.py
# Created on Mar 15, 2013

from heapq import *

class kdtree:

    # adapted from wikipedia, with slight modification
    def __init__(self, point_list, depth=0):
        if not point_list:
            self.location = None
            self.axis = 0
            self.left_child = None
            self.right_child = None
        else:
            # Select axis based on depth so that axis cycles through all valid values
            # self.axis = self.calculate_optimal_axis(point_list)
            self.axis = depth % len(point_list[0])
    
            # Sort point list and choose median as pivot element
            point_list.sort(key=lambda point: point[self.axis])
            median = len(point_list) // 2  # choose median
    
            # Create node and construct subtrees
            self.location = point_list[median]
            self.left_child = kdtree(point_list[:median], depth + 1)
            self.right_child = kdtree(point_list[median + 1:], depth + 1)

    # calculate optimal axis as the axis along which variance is maximum (Ref: http://www.slac.stanford.edu/cgi-wrap/getdoc/slac-pub-1549.pdf)
    def calculate_optimal_axis(self, point_list):
        def variance(l, dim):
            mean = 0.
            for elem in l:
                mean += elem[dim]
            mean = mean/len(l)        
            variance = 0.
            for elem in l:
                variance += (elem[dim]-mean)**2
            return variance

        optimal_dim = -1
        max_var = 0.0
        for dim in range(len(point_list[0])):
            var = variance(point_list, dim)
            if var > max_var:
                optimal_dim = dim
                max_var = var
        return optimal_dim

    def __repr__(self):
        if self.location:
            return "(%d, %s, %s, %s)" % (self.axis, repr(self.location), repr(self.left_child), repr(self.right_child))
        else:
            return "None"
        
    # keep a max-heap of k best-distances found so far. Evict the max if you find a closer point
    def knn(self, k, query, knearest=[]):
        if self.location is None:
            return knearest
 
        cur_dist = self.sqrd_distance(self.location, query)
        if len(knearest) < k:
            heappush(knearest, (-cur_dist, self.location))
        elif cur_dist < -knearest[0][0]:
            heapreplace(knearest, (-cur_dist, self.location))

        nc = self.near_child(query)
        knearest = nc.knn(k, query, knearest)
 
        max_dist = -knearest[0][0]
        if len(knearest) < k or self.axis_sqrd_distance(query) < max_dist:
            fc = self.far_child(query)
            knearest = fc.knn(k, query, knearest)
 
        return knearest

    def near_child(self, cur_point):
        if cur_point[self.axis] < self.location[self.axis]:
            return self.left_child
        else:
            return self.right_child
 
    def far_child(self, cur_point):
        if cur_point[self.axis] < self.location[self.axis]:
            return self.right_child
        else:
            return self.left_child 

    def axis_sqrd_distance(self, point):
        # axis_point = list(point)
        # axis_point[self.axis] = self.location[self.axis]
        # return self.sqrd_distance(tuple(axis_point), point)
        return (self.location[self.axis] - point[self.axis])**2

    def sqrd_distance(self, p1, p2):        
        dist = 0.
        for i in range(len(p1)):
            dist += (p1[i]-p2[i])**2
        #print "printing distance..."
        #print p1, p2, dist
        return dist

"""
num_lines_to_read = 10 #1000
input_filename = 'input.dat'
txt = open(input_filename, 'r').readlines()
points = [eval(txt[i]) for i in range(num_lines_to_read)]        
"""
# points = [(0,0), (0,1), (0,2), (1,0), (1,1), (1,2), (2,0), (2,1), (2,2)]
# points = [(0, 1), (1,0), (3,2), (2,3)]
points = [(0.2,0), (0,0.2), (0.8,0), (0,0.8)]
query_points = list(points)
kdt = kdtree(points)
# print kdt
k = 2 # one of the points will be the point itself
for i in range(len(query_points)):
    print query_points[i], kdt.knn(k, query_points[i], [])

